src/main.py

```diff
@@ -168,7 +169,50 @@ def run(options: Options) -> int:
 
         global_decls = global_info.global_decls(fmt, options.global_decls)
         if global_decls:
-            print(global_decls)
+            funcs_seen = set()
+            funcs_need_declare = set()
+            funcs_need_declare |= global_info.asm_data.mentioned_labels
+
+            for func in function_infos:
+                if not isinstance(func, FunctionInfo):
+                    continue
+
+                funcs_seen.add(func.stack_info.function.name)
+
+                for instr in func.stack_info.function.body:
+                    if not isinstance(instr, Instruction):
+                        continue
+
+                    for arg in instr.args:
+                        if isinstance(arg, AsmGlobalSymbol):
+                            func_name = arg.symbol_name
+                        elif isinstance(arg, Macro) and isinstance(
+                            arg.argument, AsmGlobalSymbol
+                        ):
+                            func_name = arg.argument.symbol_name
+                        else:
+                            continue
+
+                        if func_name in global_info.local_functions:
+                            if func_name not in funcs_seen:
+                                funcs_need_declare.add(func_name)
+
+            new_decls = []
+            for decl in global_decls.splitlines():
+                if "(" in decl and (");") in decl:
```

Comment:
String parsing is not great, and will even fail when function pointers get involved. Can we put this filtering inside global_decls instead?